### PR TITLE
spec: add locals terraform generation

### PR DESCRIPTION
**Pull request by katcipis**
Basic initial spec to how someone can leverage Terramate metadata + global definitions on Terraform code in a stack. The idea is to debate/evolve on top of this, so not final or complete, just the first idea that came from me and @i4ki talking about this.

### CODE REVIEW

**Review by mariux — CHANGES_REQUESTED** [review]
locals should not defined in context of a stack.. as this would mean code duplication in each and every stack.. it should follow same inheritance as globals to be able to leverage full power..

**Review by i4ki — APPROVED** [review]
LGTM :rocket:

**Comment by katcipis** [comment]
> locals should not defined in context of a stack.. as this would mean code duplication in each and every stack.. it should follow same inheritance as globals to be able to leverage full power..

hmm I did miss that from our talks :thinking: 

@mariux do you have any suggestions on how this would look like then ? I imagine it would be something similar to the globals block that can be anywhere then but not very sure about naming/structure, I suppose you already have something in mind.

**Comment by mariux** [comment]
the idea is to keep it DRY as much as possible.... with the current spec i see i copy 

```hcl
terramate {
  stack {
    locals {
      project     = global.project
      environment = global.environment
    }
  }
}
```

into each and every stack...

so not sure but if we would just have a 

```
locals {
  project     = global.project
  environment = global.environment
}
```

block on a higher level it would help not repeating it

same behavior as globals?

**Comment by mariux** [comment]
maybe it would even make sense to omit exporting in the stack..

```hcl
locals {
  project     = global.project
  environment = omit
}
```

this would not export `environment` anymore... so we can exclude locals from export inside a stack..

File: docs/locals-generation.md [hunk 1]
```diff
@@ -0,0 +1,188 @@
...
+```hcl
+globals {
+  data          = "data"
+  more_data     = "more data"
+  yet_more_data = "YMD"
+}
+```
+
+If we define a [configuration file](config.md) for all stacks 
+export some locals at **stacks/terramate.tm.hcl**:
```
Comment (i4ki) [inline docs/locals-generation.md:82]:
```suggestion
that export some locals at **stacks/terramate.tm.hcl**:
```

**Review by i4ki — COMMENTED** [review]
lgtm

File: docs/locals-generation.md [hunk 2]
```diff
@@ -0,0 +1,188 @@
...
+```
+
+Exporting using multiple blocks on the same [configuration file](config.md)
+is also allowed, as far as no local is redefined, the previous example could
+also be written like this:
+
+```hcl
+export_as_locals {
+  data      = globals.data
+}
```
Comment (mariux) [inline docs/locals-generation.md:38]:
```suggestion
export_as_locals {
  data = globals.data
}
```

File: docs/locals-generation.md [hunk 3]
```diff
@@ -0,0 +1,188 @@
...
+
+export_as_locals {
+  more_data = globals.more_data
```
Comment (mariux) [inline docs/locals-generation.md:41]:
```suggestion
  more_data = global.more_data
```

File: docs/locals-generation.md [hunk 4]
```diff
@@ -0,0 +1,188 @@
...
+}
+```
+
+If we define a [configuration file](config.md) for all stacks 
+that export some locals at **stacks/terramate.tm.hcl**:
+
+```hcl
+export_as_locals {
+  data      = globals.data
+  more_data = globals.more_data
```
Comment (mariux) [inline docs/locals-generation.md:87]:
```suggestion
  data      = global.data
  more_data = global.more_data
```

File: docs/locals-generation.md [hunk 5]
```diff
@@ -0,0 +1,188 @@
...
+local.data
+local.more_data
+```
+
+Now lets extend the exported locals only for **stack-1** by adding
+this to **stacks/stack-1/terramate.tm.hcl**:
+
+```hcl
+export_as_locals {
+  stack_1_only = globals.yet_more_data
```
Comment (mariux) [inline docs/locals-generation.md:124]:
```suggestion
  stack_1_only = global.yet_more_data
```

File: docs/locals-generation.md [hunk 6]
```diff
@@ -0,0 +1,188 @@
...
+}
+```
+
+To use then directly on Terraform you can instruct code generation
+to create them as locals:
+
+```hcl
+export_as_locals {
+  data      = globals.data
+  more_data = globals.more_data
```
Comment (mariux) [inline docs/locals-generation.md:27]:
```suggestion
  data      = global.data
  more_data = global.more_data
```

**Review by mariux — CHANGES_REQUESTED** [review]
typo in `global.` ;)

File: docs/locals-generation.md (hunk 4 above)
Comment (katcipis) [reply docs/locals-generation.md:87]: ow man I do this shit all the time x_x (switching global to globals :laughing: ). thanks for catching it

**Comment by katcipis** [comment]
> typo in `global.` ;)

duh

**Review by mariux — APPROVED** [review]
lgtm

**Review by mariux — APPROVED** [review]
lgtm

**Pull request merged**